#include "../include/minhajogojanela.hpp" // Nosso pr√≥prio cabe√ßalho
#include "ui_jogo.h"                  // Este arquivo √© gerado AUTOMATICAMENTE pelo CMake/uic
                                      // Ele cont√©m a defini√ß√£o da classe Ui::Form
#include <QMessageBox>                // Para mostrar mensagens pop-up
#include <QTimer>                     // Para delays e a√ß√µes futuras (como proximoTurno ap√≥s um erro)
#include <random>                     // Para gerar n√∫meros aleat√≥rios (dado)

// Construtor da nossa janela de jogo
MinhaJogoJanela::MinhaJogoJanha(QWidget *parent) :
    QMainWindow(parent), // Inicializa a parte da janela principal do Qt
    ui(new Ui::Form)     // Cria uma nova inst√¢ncia da interface gerada pelo Qt Designer
{
    ui->setupUi(this); // Configura todos os elementos da interface na nossa janela

    // Conecta os bot√µes da UI aos nossos "slots" (fun√ß√µes) C++.
    // Como renomeamos os bot√µes no Designer, o Qt j√° tenta fazer a conex√£o automaticamente.
    // Mas se quisermos ser expl√≠citos ou para garantir, podemos fazer:
    // connect(ui->optionAButton, SIGNAL(clicked()), this, SLOT(on_optionAButton_clicked()));
    // N√£o √© estritamente necess√°rio para nomes "on_ObjectName_signalName()", mas √© bom saber.

    iniciarNovoJogo(); // Come√ßa um novo jogo quando a janela √© criada
}

// Destrutor da nossa janela
MinhaJogoJanela::~MinhaJogoJanela()
{
    delete ui; // Libera a mem√≥ria da interface
}

void MinhaJogoJanela::iniciarNovoJogo() {
    jogadores = {
        {"Dom", 0},
        {"Brian", 0}
    };
    meta = 30;
    jogoTerminou = false;
    jogadorAtualIndex = 0;

    // Desabilita o bot√£o de dado no in√≠cio (s√≥ pode ser clicado ap√≥s acertar a pergunta)
    ui->rollDiceButton->setEnabled(false);
    // Habilita os bot√µes de op√ß√£o
    ui->optionAButton->setEnabled(true);
    ui->optionBButton->setEnabled(true);
    ui->optionCButton->setEnabled(true);
    ui->optionDButton->setEnabled(true);

    proximoTurno(); // Inicia o jogo com o primeiro turno
}

void MinhaJogoJanela::atualizarUI() {
    // Atualiza o texto da pergunta e das op√ß√µes nos QLabel e QPushButtons
    ui->questionLabel->setText(perguntaAtual.texto.c_str());
    ui->optionAButton->setText(perguntaAtual.opcoes[0].c_str());
    ui->optionBButton->setText(perguntaAtual.opcoes[1].c_str());
    ui->optionCButton->setText(perguntaAtual.opcoes[2].c_str());
    ui->optionDButton->setText(perguntaAtual.opcoes[3].c_str());

    // Atualiza a posi√ß√£o dos jogadores no label de posi√ß√£o
    QString posText = QString("%1: Posi√ß√£o %2/%5\n%3: Posi√ß√£o %4/%5")
                         .arg(jogadores[0].nome.c_str())
                         .arg(jogadores[0].posicao)
                         .arg(jogadores[1].nome.c_str())
                         .arg(jogadores[1].posicao)
                         .arg(meta);
    ui->playerPositionsLabel->setText(posText);
}

void MinhaJogoJanela::mostrarPergunta(const Pergunta& p) {
    perguntaAtual = p; // Armazena a pergunta atual
    atualizarUI();     // Atualiza a interface com a nova pergunta

    // Habilita os bot√µes de resposta e garante que o de dado est√° desabilitado
    ui->optionAButton->setEnabled(true);
    ui->optionBButton->setEnabled(true);
    ui->optionCButton->setEnabled(true);
    ui->optionDButton->setEnabled(true);
    ui->rollDiceButton->setEnabled(false);
}

void MinhaJogoJanela::processarResposta(char respostaUsuario) {
    // Desabilita os bot√µes de op√ß√£o para evitar cliques m√∫ltiplos
    ui->optionAButton->setEnabled(false);
    ui->optionBButton->setEnabled(false);
    ui->optionCButton->setEnabled(false);
    ui->optionDButton->setEnabled(false);

    if (respostaUsuario == perguntaAtual.correta) {
        QMessageBox::information(this, "Resposta Correta!", "Parab√©ns! Acertaste! Agora lan√ßa o dado.");
        ui->rollDiceButton->setEnabled(true); // Habilita o bot√£o de dado
    } else {
        QMessageBox::warning(this, "Resposta Errada!",
                             QString("Ops! A resposta correta era: %1. Ficas na mesma posi√ß√£o.").arg(perguntaAtual.correta));
        // Usa QTimer::singleShot para chamar proximoTurno depois de um tempo
        QTimer::singleShot(1500, this, &MinhaJogoJanela::proximoTurno);
    }
}

void MinhaJogoJanela::rolarDadoEReagir(int dado) {
    ui->rollDiceButton->setEnabled(false); // Desabilita o bot√£o de dado ap√≥s a rolagem

    Corredor& jogadorAtual = jogadores[jogadorAtualIndex];
    animarMovimentoUI(jogadorAtual, dado); // Atualiza a posi√ß√£o do jogador na UI

    if (jogadorAtual.posicao >= meta) {
        finalizarJogo(); // Se o jogador alcan√ßou ou passou a meta, o jogo termina
    } else {
        proximoTurno(); // Se n√£o, passa para o pr√≥ximo turno/jogador
    }
}

void MinhaJogoJanela::proximoTurno() {
    if (jogoTerminou) return; // N√£o faz nada se o jogo j√° terminou

    // Troca para o pr√≥ximo jogador
    jogadorAtualIndex = (jogadorAtualIndex + 1) % jogadores.size();
    Corredor& jogador = jogadores[jogadorAtualIndex];

    QMessageBox::information(this, "Vez do Jogador", QString("√â a vez de %1!").arg(jogador.nome.c_str()));

    // Obt√©m e mostra a nova pergunta para o jogador atual
    perguntaAtual = obterPerguntaAleatoria(); // Reutiliza sua fun√ß√£o de perguntas
    mostrarPergunta(perguntaAtual);
}

void MinhaJogoJanela::finalizarJogo() {
    jogoTerminou = true;
    Corredor* vencedor = &jogadores[0];
    for (auto& j : jogadores) {
        if (j.posicao > vencedor->posicao)
            vencedor = &j;
    }
    QMessageBox::information(this, "Fim de Jogo!", QString("üèÜ VENCEDOR: %1 üéâüéâüéâ").arg(vencedor->nome.c_str()));

    // Opcional: Desabilitar todos os bot√µes ou mostrar um bot√£o "Novo Jogo"
    ui->optionAButton->setEnabled(false);
    ui->optionBButton->setEnabled(false);
    ui->optionCButton->setEnabled(false);
    ui->optionDButton->setEnabled(false);
    ui->rollDiceButton->setEnabled(false);
}

void MinhaJogoJanela::animarMovimentoUI(Corredor& jogador, int dado) {
    // Para uma anima√ß√£o visual mais complexa, voc√™ precisaria de mais trabalho (QGraphicsView, QPropertyAnimation).
    // Por enquanto, apenas atualiza a posi√ß√£o e mostra uma mensagem.
    jogador.posicao += dado;
    atualizarUI(); // Atualiza a posi√ß√£o na interface imediatamente
}

// --- Slots que s√£o chamados quando os bot√µes s√£o clicados ---

void MinhaJogoJanela::on_optionAButton_clicked() {
    processarResposta('A');
}

void MinhaJogoJanela::on_optionBButton_clicked() {
    processarResposta('B');
}

void MinhaJogoJanela::on_optionCButton_clicked() {
    processarResposta('C');
}

void MinhaJogoJanela::on_optionDButton_clicked() {
    processarResposta('D');
}

void MinhaJogoJanela::on_rollDiceButton_clicked() {
    // Simula a rolagem do dado
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(1, 6);
    int dado = distrib(gen);

    QMessageBox::information(this, "Rolagem do Dado", QString("üé≤ Saiu: %1!").arg(dado));
    rolarDadoEReagir(dado);
}